dafni/src/main_dafni.py

```diff
@@ -91,53 +84,33 @@ def get_args(test_args=None) -> argparse.Namespace:
     return args
 
 
-def read_variables(variables_path: Path) -> FileNotFoundError | dict:
+def parse_variables(causal_dag: CausalDAG) -> tuple:
     """
-    Function to read the variables.json file specified by the user
-    :param variables_path: A Path object of the user-specified file path
+    Function to parse and validate the variables defined in the causal dag.
+    :param causal_dag: an instancce of the CausalDAG class containing the relationships and variables (as attribtutes)
     :returns:
-            - dict - A valid dictionary consisting of the causal tests
+    - Tuple containing the inputs, outputs and constraints to pass into the modelling scenario
     """
-    if not variables_path.exists() or variables_path.is_dir():
-        print(f"JSON file not found at the specified location: {variables_path}")
 
-        raise FileNotFoundError
+    inputs = [
+        Input(node, eval(eval(attributes["datatype"])))
+        for node, attributes in causal_dag.graph.nodes(data=True)
+        if eval(attributes["typestring"]) == "input"
+    ]
 
-    with variables_path.open("r") as file:
-        inputs = json.load(file)
+    constraints = set()
 
-    return inputs
+    for (node, attributes), input_var in zip(causal_dag.graph.nodes(data=True), inputs):
```

Comment:
Expressing constraints in terms of individual variables is a real headache. I battled with this for about 6 months during my PhD. The problem is that constraints can relate variables, e.g. `i1 > i2`. Then you need to put the constraint `i1 > i2` for node `i1` and the inverse (i.e. `i2 < i1`) for node `i2`. (I guess you don't technically need to do this here, but then you have the question of which node you associate the constraint with.)

Reply:
When I did the [CARLA case study](https://github.com/CITCOM-project/carla-case-study/blob/main/studied-cases/case-1/causal_tests.json) I encoded queries on a per-test basis using an attribute called "query", which is a silly name for it here but made sense at the time because you can then just do `df = df.query(test["query"])` and it'll "just work" if the user supplies the query in the correct format. I propose we do a similar thing here.